scapy/contrib/automotive/xcp/xcp.py

```diff
@@ -86,8 +87,10 @@ class XCPOnCAN(CAN):
     ]
 
     def post_build(self, pkt, pay):
-        if self.length is None:
-            tmp_len = len(pay)
+        if self.length is None or \
+                (len(pay) < 8 and conf.contribs["XCP"]["add_padding_for_can"]):
+            tmp_len = 8 if conf.contribs["XCP"]["add_padding_for_can"] else \
+                len(pay)
             pkt = pkt[:4] + struct.pack("B", tmp_len) + pkt[5:]
```

Comment:
I think it's better to modify the payload itself than just setting the length parameter.
See also https://github.com/secdev/scapy/pull/1595#pullrequestreview-156300314
And here you can find how it's done for ISOTP https://github.com/secdev/scapy/pull/2963/commits/6620501358d189bca7b4d8264d17c376bb0c505c
Also a unit test would be nice to test the padding.

@polybassa Do you have some knowledge about how often padding is used for CAN communication? Maybe it's worth it to move this functionality to the CANSocket itself like for the ISOTP sockets.
Then `conf.contribs["XCP"]["add_padding_for_can"]` wouldn't be necessary too.

Reply:
I would keep padding protocol specific. It really depends on the manufacturer of the ECU.